man/devas-visualize-geometry.1
.TH DEVA-VISUALIZE-GEOMETRY 1 "22 January 2018" "DEVA Project"
.SH NAME
devas-visualize-geometry \- generate displayable DEVA geometry file
.SH SYNOPSIS
.TP
\fBdevas-visualize-geometry\fR [\fIoptions\fR] [\fIprint-row-start print-col-start
print-row-end print-col-end\fR] \fIinput.txt\fR \fIoutput.pdf\fR
.SH DESCRIPTION
Generate displayable version of DEVA geometry file as a PNG file.  3D
values are displayed in RGB, with R=|x|, G=|y|, and B=|z|.  1D values
are displayed with positive values in green and negative values in red.
Black indicates a zero value.  Display values are scaled by the extremal
value of the input.  For 1d data, effectively infinite values (holes in
distance data) are displayed in orange.  This is not needed for xyz or
nor data, since in those data files hole result in zero values.
.SH OPTIONS
.TP
\fB\-\-3dx1=\fIa\fR
Display a single dimension of at 3D file using the same coloring scheme
as for 1D data.  \fIa\fR is a single characters specifying the dimension
('x', 'y', or 'z').
.TP
\fB\-\-print\fR
Print selected values in the input to standard output.  Requires four
additional parameters:
\fIprint-row-start\fR,
\fIprint-col-start\fR,
\fIprint-row-end\fR, and
\fIprint-col-end\fR, specifying the upper-left and lower-right corners
of the values to be printed.
.TP
\fB\-\-outline\fR
Outline the location of the printed values in the output PNG file.
.SH EXAMPLES
To visualize dst (distance from viewpoint) data:
.IP "" .5i
devas-visualize-geometry input_dst.txt output_dst.png
.PP
To visualize the Y dimension of an xyz.txt (surface point location) file:
.IP "" .5i
devas-visualize-geometry --3dx1=y input_xyz.txt output_xyz.png
\." SH LIMITATIONS
\." PP
.SH AUTHOR
William B. Thompson


CHANGES
version 4.0.00

Updated MacOS and Windows binaries to current versions of source.

Renamed files and code objects from "DIVA" to "DeVAS" to avoid conflicts
with DIVA for RHINO.

version 3.3.04

Added comment to deva-margin.c explaining how the margin pixels were
computed.

version 3.3.03

Print version number on -v or --v.  Previously, needed  -version or
--version.

version 3.3.02

Issue --legalblind prompt when executing deva-visibility with no
arguments.

version 3.3.01

Better error reporting in read-geometry for Radiance .txt file with no
empty line terminating header.

version 3.3.00

Most fatal errors now print out the line number and file where the error
occurred.

Added option in deva-visibility to check for false positive luminance
boundary detections.

Added feature in deva-visibility (deva-commandline) and
deva-compare-boundaries to check for very dark regions in original
image.  This is motivated by a lack of knowledge about the effects of
illumination levels below about 1 cd/m^2 on low vision acuity and
contrast sensitivity.

FFT-based Gaussian blur routine added (deva-gblur-fft). This is used by
the code that checks for very dark regions.

Added routines in deva-margin for DEVA_float image object, providing
support needed by code that finds very dark regions in original image.

Added code in visualize-hazards for displaying very dark regions and
false positive errors in hazard visualization.

Added code supporting region-of-interest (ROI) processing in
determination of Visibility Hazard Scores in deva-visibility.

Changed contrast sensitivity of presets based on personal communication
(G.E. Legge & Y-Z Xiong, Feb. 19, 2019).

Added a "legalblind" preset to the other low vision categories in
deva-filter and deva-visibility.  This specifies the U.S.
"legal blindness" acuity level along with appropriate contrast
sensitivity and color sensitivity values.  Note that this specifies a
level of visual impairment right at the legal blindness cutoff point.
It is not some sort of test of legal blindness.

Added a "normal" preset to the other low vision categories in
deva-filter and deva-visibility.  This is treated as a silent option,
not included in the prompt or man page, because the filter has not been
tested for this condition.

Cleaned up where new image objects were allocated in deva-visibility.c
and related routines.

Minor comment cleanup.

Cleanups and additions to man pages.

Various changes to CMakeLists.txt

Ran this file through a spell checker.

version 3.2.04

luminance-boundaries now issues a warning to avoid use with input images
generated by deva-filter.  The problem may be due to quantization noise
in the low precision Radiance format input image resulting in a large
number of micro edges that in turn confuse the auto thresholding
routine.

version 3.2.03

Really fix deva-visibility man page.

version 3.2.02

Minor fix to deva-visibility man page.

version 3.2.01

Added build process for MacOS and standardized directory layout for
Windows and MacOS builds.

Removed extraneous debugging command from deva-margin.c

version 3.2.00

Added two new programs: luminance-boundaries, geometry-boundaries.

In deva-filter.c (deva-filter, deva-visibility): Changed smoothing
algorithm to better deal with artifacts at the boundaries of the
smoothing windows.  Added optional debugging code to output individual
contrast bands.

In deva-margin.c:  Added compile-time option to base feathering on
average luminance over full image rather than just over the boundary
pixels.  Added compile-time option to base margin pixels on reflection
of image around the boundary, not just a fill of boundary pixels.

In dilate.c, added dt_euclid_sq_2 which used a pre-allocated output
image.

In deva-commandline.c (deva-filter, deva-visibility): Cleaned up command
summaries.  Allowed specifying smoothing_type with presets.

In ChungLeggeCSF.h: Cleanup of unused ChungLeggeCSF_PEAK_FREQUENCY
defines.

Man pages for new programs, minor updates to existing man pages.

version 3.1.02

In deva-commandline.c (deva-visibility): Added measurement type options
to usage message.

version 3.1.01

deva-compare-boundaries.c: New program to create a visualization image
comparing a standard set of boundary elements (e.g., geometry) to a
different set of boundary elements (e.g., luminance edges).  Annotation
of output image, which requires Cairo library, is includable with a
compile-time switch.

deva-add-text.c: New routine to add annotation text to an image.
Requires a switch in CMakeLists.txt to link against Cairo library.

In deva-commandline.c (deva-filter, deva-visibility): Switch to basing
auto-clipping on multiple of median luminance, which is more robust than
the previous multiple of average luminance.  Added a compile-time option
for basing margin on only smaller dimension of input image.

In deva-commandline.c (deva-visibility): Added interface for more
measurement types.

In deva-canny.c:  Switch to basing luminance boundaries on log
luminance, not luminance itself.  This better approximates human
perception of contrast boundaries.  Switched to basing auto-thresholding
on local maxima of directional gradient, not all directional gradients.

In visualize-hazards.c/visualize-hazards.h: Cleaned up calling sequence.
Added in more measurement types.
 
In read-geometry.c, fixed possibly uninitialized DEVA_null_view.

Additional commenting.

version 3.0.01

Modified radiance-header.c to deal with indented VIEW records produced
by pcomb.  (pcomb has been modified in development version of Radiance
to output one non-indented VIEW record, alleviating the need for this
change if a recent enough version of Radiance is used to generate the
file.)

Cleaned up deva-gblur.c.  There should be no change in functionality.

Better error reporting in read-geometry routines.


devas-visualize-geometry.c
/*
 * Generate displayable version of DeVAS geometry file as a PNG file.
 * 3D values are displayed in RGB, with R=|x|, G=|y|, and B=|z|.  1D
 * values are displayed with positive values in green and negative
 * values in red.  Black indicates a zero value.  The --3dx1=dim flag
 * displays a single dimension of at 3D file using the same coloring
 * scheme as for 1D data.  <a> is a single characters specifying the
 * dimension ('x', 'y', or 'z').  For example, the following could be
 * used for displaying the Y dimension of an xyz.txt file:
 *
 * 	devas-visualize-geometry --3dx1=y input_xyz.txt output.png
 *
 * Display values are scaled by the extremal value of the input.  For
 * 1d data, effectively infinite values (holes in distance data) are
 * displayed in orange.  This is not needed for xyz or nor data, since
 * in those data files hole result in zero values.
 */

/* #define	DeVAS_CHECK_BOUNDS */

#define	MAX_DST		(1e+8)	/* any distance larger than this is infinite */
				/* and thus a hole */
#define	WARN_RED	255
#define	WARN_GREEN	127
#define	WARN_BLUE	0

#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <strings.h>
#include <string.h>
#include <ctype.h>
#include "devas-image.h"
#include "read-geometry.h"
#include "devas-sRGB.h"
#include "devas-png.h"
#include "devas-filter-version.h"
#include "devas-license.h"

char	*Usage =
	    "devas-visualize-geometry [--3dx1=<a>] [--print] [--outline]"
	    "\n\t[print-row-start print-col-start print-row-end print-col-end]"
	    "\n\tgeom.txt vis.png";
int	args_needed = 2;

double  rescale ( double old_value, double old_min, double old_max,
	    double new_min, double new_max );
int	print_coordinates_in_image ( int print_row_start, int print_col_start,
	    int print_row_end, int print_col_end, int n_rows, int n_cols );

int
main ( int argc, char *argv[] )
{
    char		dim_to_show = 'n';	/* default is to show 3D */
    /* data with each dimension */
    /* in a different color */
    int			geo_dim;
    DeVAS_float_image	*data_1d;
    DeVAS_XYZ_image	*data_3d;
    DeVAS_RGB_image	*display;
    DeVAS_RGBf		display_value_float;
    int			row, col;
    float		max_X, max_Y, max_Z, max_1d;
    float		min_X, min_Y, min_Z, min_1d;
    float		max_all_dimensions, min_all_dimensions;
    float		norm_3d, norm_1d;
    int			verbose = FALSE;
    int			warn_only = FALSE;
    int			fullrange = FALSE;
    int			reverse = FALSE;
    int			found_valid_point;
    int			print_flag = FALSE;
    int			print_row_start = -1;
    int			print_col_start = -1;
    int			print_row_end = -1;
    int			print_col_end = -1;
    int			outline_flag = FALSE;
    DeVAS_RGB		cyan = { 0, 255, 255 };

    int			argpt = 1;

    /* scan for flags */
    while ( ( ( argc - argpt ) >= 1 ) && ( argv[argpt][0] == '-' ) ) {
	if ( strncasecmp ( argv[argpt], "--3dx1=",
		    strlen ( "--3dx1=" ) ) == 0 ) {
	    dim_to_show = tolower ( *( argv[argpt] + strlen ( "--3dx1=" ) ) );
	    argpt++;
	} else if ( strncasecmp ( argv[argpt], "-3dx1=",
		    strlen ( "-3dx1=" ) ) == 0 ) {
	    dim_to_show = *( argv[argpt] + strlen ( "-3dx1=" ) );
	    argpt++;
	} else if ( ( strcasecmp ( argv[argpt], "--fullrange" ) == 0 ) ||
		( strcasecmp ( argv[argpt], "-fullrange" ) == 0 ) ) {
	    fullrange = TRUE;
	    argpt++;
	} else if ( ( strcasecmp ( argv[argpt], "--reverse" ) == 0 ) ||
		( strcasecmp ( argv[argpt], "-reverse" ) == 0 ) ) {
	    reverse = TRUE;
	    fullrange = TRUE;
	    argpt++;
	} else if ( ( strcasecmp ( argv[argpt], "--warnonly" ) == 0 ) ||
		( strcasecmp ( argv[argpt], "-warnonly" ) == 0 ) ) {
	    warn_only = TRUE;
	    argpt++;
	} else if ( ( strcasecmp ( argv[argpt], "--print" ) == 0 ) ||
		( strcasecmp ( argv[argpt], "-print" ) == 0 ) ) {
	    print_flag = TRUE;
	    args_needed += 4;
	    argpt++;
	} else if ( ( strcasecmp ( argv[argpt], "--outline" ) == 0 ) ||
		( strcasecmp ( argv[argpt], "-outline" ) == 0 ) ) {
	    outline_flag = TRUE;
	    argpt++;
	} else if ( ( strcasecmp ( argv[argpt], "--verbose" ) == 0 ) ||
		( strcasecmp ( argv[argpt], "-verbose" ) == 0 ) ) {
	    verbose = TRUE;
	    argpt++;
	} else {
	    fprintf ( stderr, "invalid flag (%s)!\n", argv[argpt] );
	    return ( EXIT_FAILURE );    /* error exit */
	}
    }

    /* check number of arguments */
    if ( ( argc - argpt ) != args_needed ) {
	fprintf ( stderr, "%s\n", Usage );
	return ( EXIT_FAILURE );	/* error return */
    }

    /* other error checks */
    if ( ( dim_to_show != 'n' ) && ( dim_to_show != 'x' ) &&
	    ( dim_to_show != 'y' ) && ( dim_to_show != 'z' ) )  {
	fprintf ( stderr, "invalid dim_to_show (%d)!\n", dim_to_show );
	DeVAS_print_file_lineno ( __FILE__, __LINE__ );
	return ( EXIT_FAILURE );    /* error exit */
    }

    if ( outline_flag && !print_flag ) {
	fprintf ( stderr, "invalid dim_to_show (%d)!\n", dim_to_show );
	DeVAS_print_file_lineno ( __FILE__, __LINE__ );
	return ( EXIT_FAILURE );    /* error exit */
    }

    if ( print_flag ) {
	print_row_start = atoi ( argv[argpt++] );
	print_col_start = atoi ( argv[argpt++] );
	print_row_end = atoi ( argv[argpt++] );
	print_col_end = atoi ( argv[argpt++] );
    }

    geo_dim = DeVAS_geom_dim_from_radfilename ( argv[argpt] );
    /* need to reuse this argument below */

    /* argument compatibility check */
    if ( ( geo_dim == 1 ) && ( dim_to_show != 'n' ) ) {
	fprintf ( stderr, "can't specify --3dx1 for 1D data!\n" );
	DeVAS_print_file_lineno ( __FILE__, __LINE__ );
	return ( EXIT_FAILURE );    /* error exit */
    }

    if ( geo_dim == 1 ) {
	/* input is 1D */

	data_1d = DeVAS_geom1d_from_radfilename ( argv[argpt++] );

	if ( print_flag ) {
	    if ( print_coordinates_in_image ( print_row_start, print_col_start,
			print_row_end, print_col_end,
			DeVAS_image_n_rows ( data_1d ),
			DeVAS_image_n_cols ( data_1d ) ) ) {
		fprintf ( stderr,
			"print points outsize image or otherwise invalid!\n" );
		DeVAS_print_file_lineno ( __FILE__, __LINE__ );
		exit ( EXIT_FAILURE );
	    }

	    for ( row = print_row_start; row <= print_row_end; row++ ) {
		for ( col = print_col_start; col <= print_col_end; col++ ) {
		    printf ( "%f ", DeVAS_image_data ( data_1d, row, col ) );
		}
		printf ( "\n" );
	    }
	    printf ( "\n" );
	}

	display = DeVAS_RGB_image_new ( DeVAS_image_n_rows ( data_1d ),
		DeVAS_image_n_cols ( data_1d ) );

	/* find extremal value in input */
	max_1d = min_1d = 0.0;	/* quiet -Wunused-variable */
	found_valid_point = FALSE;
	for ( row = 0; row < DeVAS_image_n_rows ( data_1d ); row++ ) {
	    for ( col = 0; col < DeVAS_image_n_cols ( data_1d ); col++ ) {
		if ( DeVAS_image_data ( data_1d, row, col ) < MAX_DST ) {
		    max_1d = min_1d = DeVAS_image_data ( data_1d, row, col );
		    found_valid_point = TRUE;
		}
	    }
	}
	if ( !found_valid_point ) {
	    fprintf ( stderr, "no valid data!\n" );
	    DeVAS_print_file_lineno ( __FILE__, __LINE__ );
	    exit ( EXIT_FAILURE );
	}

	for ( row = 0; row < DeVAS_image_n_rows ( data_1d ); row++ ) {
	    for ( col = 0; col < DeVAS_image_n_cols ( data_1d ); col++ ) {
		if ( DeVAS_image_data ( data_1d, row, col ) > MAX_DST ) {
		    continue;
		}
		if ( max_1d < DeVAS_image_data ( data_1d, row, col ) ) {
		    max_1d = DeVAS_image_data ( data_1d, row, col );
		} else if ( min_1d > DeVAS_image_data ( data_1d, row, col ) ) {
		    min_1d = DeVAS_image_data ( data_1d, row, col );
		}
	    }
	}

	if ( verbose ) {
	    printf ( "min_1d = %f, max_1d = %f\n", min_1d, max_1d );
	}

	if ( ( max_1d - min_1d ) == 0 ) {
	    fprintf ( stderr, "no dynamic range in geometry file!\n" );
	    DeVAS_print_file_lineno ( __FILE__, __LINE__ );
	    return ( EXIT_FAILURE );	/* error exit */
	}

	/* create visualization image, normalizing as appropriate */

	norm_1d = 1.0 / fmax ( max_1d, -min_1d );

	for ( row = 0; row < DeVAS_image_n_rows ( data_1d ); row++ ) {
	    for ( col = 0; col < DeVAS_image_n_cols ( data_1d ); col++ ) {
		if ( DeVAS_image_data ( data_1d, row, col ) <= MAX_DST ) {
		    if ( warn_only ) {
			display_value_float.red = 0.0;
			display_value_float.green = 0.0;
			display_value_float.blue = 0.0;
		    } else {
			if ( fullrange && !reverse ) {
			    display_value_float.red =
				display_value_float.green =
				display_value_float.blue =
				rescale ( DeVAS_image_data ( data_1d, row,
					    col ), min_1d, max_1d,
					0.0, 1.0 );

			} else if ( fullrange && reverse ) {

			    display_value_float.red =
				display_value_float.green =
				display_value_float.blue = 1.0 -
				rescale ( DeVAS_image_data ( data_1d, row,
					    col ), min_1d, max_1d,
					0.0, 1.0 );
			} else if ( DeVAS_image_data ( data_1d, row, col )
				< 0.0 ) {
			    display_value_float.red = norm_1d *
				-DeVAS_image_data ( data_1d, row, col );
			    display_value_float.green = 0.0;
			    display_value_float.blue = 0.0;
			} else {
			    display_value_float.red = 0.0;
			    display_value_float.green = norm_1d *
				DeVAS_image_data ( data_1d, row, col );
			    display_value_float.blue = 0.0;
			}
		    }
		} else {
		    display_value_float.red = WARN_RED;
		    display_value_float.green = WARN_GREEN;
		    display_value_float.blue = WARN_BLUE;
		}

		DeVAS_image_data ( display, row, col ) =
		    /* RGBf_to_sRGB ( display_value_float ); */
		    RGBf_to_RGB ( display_value_float );
		/* perceptual encoding */
	    }
	}

	DeVAS_float_image_delete ( data_1d );

    } else if ( ( geo_dim == 3 ) && ( dim_to_show == 'n' ) ) {
	/*
	 * 3-D data, show absolute value of each dimension in a different
	 * color.
	 */

	if ( fullrange ) {
	    fprintf ( stderr, "--fullrange not valid for 3D data!\n" );
	    DeVAS_print_file_lineno ( __FILE__, __LINE__ );
	    exit ( EXIT_FAILURE );
	}

	if ( reverse ) {
	    fprintf ( stderr, "--reverse not valid for 3D data!\n" );
	    DeVAS_print_file_lineno ( __FILE__, __LINE__ );
	    exit ( EXIT_FAILURE );
	}

	data_3d = DeVAS_geom3d_from_radfilename ( argv[argpt++] );

	if ( print_flag ) {
	    if ( print_coordinates_in_image ( print_row_start, print_col_start,
			print_row_end, print_col_end,
			DeVAS_image_n_rows ( data_3d ),
			DeVAS_image_n_cols ( data_3d ) ) ) {
		fprintf ( stderr,
			"print points outsize image or otherwise invalid!\n" );
		DeVAS_print_file_lineno ( __FILE__, __LINE__ );
		exit ( EXIT_FAILURE );
	    }

	    for ( row = print_row_start; row <= print_row_end; row++ ) {
		for ( col = print_col_start; col <= print_col_end; col++ ) {
		    printf ( "(%.2f,%.2f,%.2f) ",
			    DeVAS_image_data ( data_3d, row, col ) . X,
			    DeVAS_image_data ( data_3d, row, col ) . Y,
			    DeVAS_image_data ( data_3d, row, col ) . Z );
		}
		printf ( "\n" );
	    }
	    printf ( "\n" );
	}

	display = DeVAS_RGB_image_new ( DeVAS_image_n_rows ( data_3d ),
		DeVAS_image_n_cols ( data_3d ) );

	/* find extremal value in input */

	max_X = min_X = DeVAS_image_data ( data_3d, 0, 0 ) . X;
	max_Y = min_Y = DeVAS_image_data ( data_3d, 0, 0 ) . Y;
	max_Z = min_Z = DeVAS_image_data ( data_3d, 0, 0 ) . Z;

	for ( row = 0; row < DeVAS_image_n_rows ( data_3d ); row++ ) {
	    for ( col = 0; col < DeVAS_image_n_cols ( data_3d ); col++ ) {

		if ( max_X <
			DeVAS_image_data ( data_3d, row, col ) . X ) {
		    max_X = DeVAS_image_data ( data_3d, row, col ) . X;
		} else if ( min_X >
			DeVAS_image_data ( data_3d, row, col ) . X ) {
		    min_X = DeVAS_image_data ( data_3d, row, col ) . X;
		}

		if ( max_Y <
			DeVAS_image_data ( data_3d, row, col ) . Y ) {
		    max_Y = DeVAS_image_data ( data_3d, row, col ) . Y;
		} else if ( min_Y >
			DeVAS_image_data ( data_3d, row, col ) . Y ) {
		    min_Y = DeVAS_image_data ( data_3d, row, col ) . Y;
		}

		if ( max_Z <
			DeVAS_image_data ( data_3d, row, col ) . Z ) {
		    max_Z = DeVAS_image_data ( data_3d, row, col ) . Z;
		} else if ( min_Z >
			DeVAS_image_data ( data_3d, row, col ) . Z ) {
		    min_Z = DeVAS_image_data ( data_3d, row, col ) . Z;
		}
	    }
	}

	if ( verbose ) {
	    printf ( "min_X = %f, max_X = %f\n", min_X, max_X );
	    printf ( "min_Y = %f, max_Y = %f\n", min_Y, max_Y );
	    printf ( "min_Z = %f, max_Z = %f\n", min_Z, max_Z );
	}

	max_all_dimensions = fmax ( fmax ( max_X, max_Y ), max_Z );
	min_all_dimensions = fmin ( fmin ( min_X, min_Y ), min_Z );

	if ( ( max_all_dimensions - min_all_dimensions ) == 0 ) {
	    fprintf ( stderr,
		    "no dynamic range in geometry file!\n" );
	    DeVAS_print_file_lineno ( __FILE__, __LINE__ );
	    return ( EXIT_FAILURE );	/* error exit */
	}

	/* create visualization image, normalizing as appropriate */

	norm_3d = 1.0 / fmax ( max_all_dimensions, -min_all_dimensions );

	for ( row = 0; row < DeVAS_image_n_rows ( data_3d ); row++ ) {
	    for ( col = 0; col < DeVAS_image_n_cols ( data_3d ); col++ ) {
		display_value_float.red = fabs ( norm_3d *
			DeVAS_image_data ( data_3d, row, col ) . X );
		display_value_float.green = fabs ( norm_3d *
			DeVAS_image_data ( data_3d, row, col ) . Y );
		display_value_float.blue = fabs ( norm_3d *
			DeVAS_image_data ( data_3d, row, col ) . Z );

		DeVAS_image_data ( display, row, col ) =
		    /* RGBf_to_sRGB ( display_value_float ); */
		    RGBf_to_RGB ( display_value_float );
	    }
	}

	DeVAS_XYZ_image_delete ( data_3d );

    } else if ( ( geo_dim == 3 ) && ( dim_to_show != 'n' ) ) {
	/* 3-D data, show selected dimension as for 1-D data */

	data_3d = DeVAS_geom3d_from_radfilename ( argv[argpt++] );
	display = DeVAS_RGB_image_new ( DeVAS_image_n_rows ( data_3d ),
		DeVAS_image_n_cols ( data_3d ) );

	data_1d = DeVAS_float_image_new ( DeVAS_image_n_rows ( data_3d ),
		DeVAS_image_n_cols ( data_3d ) );

	/* copy selected dimension to 1-D image */
	switch ( dim_to_show ) {
	    case 'x':
		for ( row = 0; row < DeVAS_image_n_rows ( data_3d ); row++ ) {
		    for ( col = 0; col < DeVAS_image_n_cols ( data_3d );
			    col++ ) {
			DeVAS_image_data ( data_1d, row, col ) =
			    DeVAS_image_data ( data_3d, row, col ) . X;
		    }
		}
		break;

	    case 'y':
		for ( row = 0; row < DeVAS_image_n_rows ( data_3d ); row++ ) {
		    for ( col = 0; col < DeVAS_image_n_cols ( data_3d );
			    col++ ) {
			DeVAS_image_data ( data_1d, row, col ) =
			    DeVAS_image_data ( data_3d, row, col ) . Y;
		    }
		}
		break;

	    case 'z':
		for ( row = 0; row < DeVAS_image_n_rows ( data_3d ); row++ ) {
		    for ( col = 0; col < DeVAS_image_n_cols ( data_3d );
			    col++ ) {
			DeVAS_image_data ( data_1d, row, col ) =
			    DeVAS_image_data ( data_3d, row, col ) . Z;
		    }
		}
		break;

	    default:
		fprintf ( stderr, "internal error!\n" );
		DeVAS_print_file_lineno ( __FILE__, __LINE__ );
		return ( EXIT_FAILURE );    /* error exit */
		break;
	}

	if ( print_flag ) {
	    if ( print_coordinates_in_image ( print_row_start, print_col_start,
			print_row_end, print_col_end,
			DeVAS_image_n_rows ( data_1d ),
			DeVAS_image_n_cols ( data_1d ) ) ) {
		fprintf ( stderr,
			"print points outsize image or otherwise invalid!\n" );
		DeVAS_print_file_lineno ( __FILE__, __LINE__ );
		exit ( EXIT_FAILURE );
	    }

	    for ( row = print_row_start; row <= print_row_end; row++ ) {
		for ( col = print_col_end; col <= print_col_end; col++ ) {
		    printf ( "%f ", DeVAS_image_data ( data_1d, row, col ) );
		}
		printf ( "\n" );
	    }
	    printf ( "\n" );
	}

	/* find extremal value in input */

	max_1d = min_1d = DeVAS_image_data ( data_1d, 0, 0 );

	for ( row = 0; row < DeVAS_image_n_rows ( data_1d ); row++ ) {
	    for ( col = 0; col < DeVAS_image_n_cols ( data_1d ); col++ ) {
		if ( max_1d < DeVAS_image_data ( data_1d, row, col ) ) {
		    max_1d = DeVAS_image_data ( data_1d, row, col );
		} else if ( min_1d > DeVAS_image_data ( data_1d, row, col ) ) {
		    min_1d = DeVAS_image_data ( data_1d, row, col );
		}
	    }
	}

	if ( verbose ) {
	    printf ( "min_%c = %f, max_%c = %f\n", dim_to_show, min_1d,
		    dim_to_show, max_1d );
	}

	if ( ( max_1d - min_1d ) == 0 ) {
	    fprintf ( stderr,
		    "no dynamic range in geometry file!\n" );
	    DeVAS_print_file_lineno ( __FILE__, __LINE__ );
	    return ( EXIT_FAILURE );	/* error exit */
	}

	/* create visualization image, normalizing as appropriate */

	norm_1d = 1.0 / fmax ( max_1d, -min_1d );

	for ( row = 0; row < DeVAS_image_n_rows ( data_1d ); row++ ) {
	    for ( col = 0; col < DeVAS_image_n_cols ( data_1d ); col++ ) {
		if ( DeVAS_image_data ( data_1d, row, col ) < 0.0 ) {
		    display_value_float.red = norm_1d *
			-DeVAS_image_data ( data_1d, row, col );
		    display_value_float.green = 0.0;
		    display_value_float.blue = 0.0;
		} else {
		    display_value_float.red = 0.0;
		    display_value_float.green = norm_1d *
			DeVAS_image_data ( data_1d, row, col );
		    display_value_float.blue = 0.0;
		}
		DeVAS_image_data ( display, row, col ) =
		    RGBf_to_sRGB ( display_value_float );
	    }
	}

	DeVAS_float_image_delete ( data_1d );
	DeVAS_XYZ_image_delete ( data_3d );

    } else {
	fprintf ( stderr, "internal error!\n" );
	DeVAS_print_file_lineno ( __FILE__, __LINE__ );
	return ( EXIT_FAILURE );    /* error exit */
    }

    if ( outline_flag ) {
	for ( row = print_row_start; row <= print_row_end; row ++ ) {
	    DeVAS_image_data ( display, row, print_col_start ) =
		DeVAS_image_data ( display, row, print_col_end ) = cyan;
	}

	for ( col = print_col_start; col <= print_col_end; col ++ ) {
	    DeVAS_image_data ( display, print_row_start, col ) =
		DeVAS_image_data ( display, print_row_end, col ) = cyan;
	}
    }

    DeVAS_RGB_image_to_filename_png ( argv[argpt++], display );

    DeVAS_RGB_image_delete ( display );

    return ( EXIT_SUCCESS );	/* normal exit */
}

double
rescale ( double old_value, double old_min, double old_max, double new_min,
	        double new_max )
/*
 * Rescale old_value, asserted to be in range [old_min - old_max] to new
 * range [new_min - new_max].
 */
{
    double new_value;

    new_value = ( ( ( old_value - old_min ) *
		    ( new_max - new_min ) ) / ( old_max - old_min ) ) +
			new_min;

    return ( new_value );
}

int
print_coordinates_in_image ( int print_row_start, int print_col_start,
	int print_row_end, int print_col_end, int n_rows, int n_cols )
{
    return ( !( ( print_row_start >= 0 ) && ( print_row_start < n_rows ) &&
	    ( print_row_end >= 0 ) && ( print_row_end < n_rows ) &&
        ( print_col_start >= 0 ) && ( print_col_start < n_cols ) &&
	    ( print_col_end >= 0 ) && ( print_col_end < n_cols ) &&
	( print_row_start <= print_row_end ) &&
	    ( print_col_start <= print_col_end ) ) );
}
